File: solutions/python/grade-school/1/grade_school.py
class School:
    def __init__(self):
        self.students = {}
        self._added_status = []

    def add_student(self, name, grade):
        if name in self.students:
            self._added_status.append(False)
            return
        self.students[name] = grade
        self._added_status.append(True)

    def roster(self):
        sorted_names = sorted(self.students.keys(), key=lambda x: (self.students[x], x))
        return sorted_names

    def grade(self, grade_number):
        return sorted([name for name, g in self.students.items() if g == grade_number])

    def added(self):
        return self._added_status

File: solutions/ruby/grade-school/1/grade_school.rb
class School
  def initialize
    @roster = {}
    @students = {}
  end

  def roster
    @roster.keys.sort.flat_map { |g| @roster[g] }
  end

  def add(name, grade)
    return false if @students.key?(name)
    @students[name] = grade
    @roster[grade] ||= []
    @roster[grade] << name
    @roster[grade].sort!
    true
  end

  def grade(n)
    @roster[n] || []
  end
end
